File: packages/ml-inference/src/services/export/acr_plaintext_renderer.py
"""acr_plaintext_renderer — Python twin of TS acrPlainTextRenderer.

Implements ``specs/002-acr-structured-readout/contracts/plaintext-renderer.md``.
Output is byte-equivalent to ``packages/app/src/emr/services/report/acrPlainTextRenderer.ts``
for the same ``ReadoutSnapshot`` input. The cross-channel parity test
(``tests/integration/test_acr_renderer_cross_channel_parity.py``) is the
release-blocking gate that enforces this contract.
"""
from __future__ import annotations

import unicodedata
from typing import Any, Iterable, Mapping, Sequence


def render_readout_plain_text(snapshot: Mapping[str, Any]) -> str:
    """Convert a ReadoutSnapshot dict into plain text.

    Snapshot dict shape mirrors the TS ``ReadoutSnapshot`` interface:
    ``{analysisId, tenantId, locale, capturedAt, etag, status,
       sections: [{section, title, rows: [{label, value, warning, itemId,
       segment, interpretation, badge, stale, key}], status, emptyMessage}],
       ruoDisclaimer}``.
    """
    raw_disclaimer = str(snapshot.get("ruoDisclaimer", "")).strip()
    # Match TS renderer normalisation: strip any leading/trailing dashes
    # the caller may have already inserted so the banner is always
    # exactly `--- <text> ---`.
    stripped = raw_disclaimer
    while stripped.startswith("---"):
        stripped = stripped[3:].lstrip()
    while stripped.endswith("---"):
        stripped = stripped[:-3].rstrip()
    ruo_banner = f"--- {stripped} ---"

    lines: list[str] = [ruo_banner, ""]
    sections: Sequence[Mapping[str, Any]] = snapshot.get("sections", []) or []
    for idx, section in enumerate(sections):
        if idx > 0:
            lines.append("")
        _render_section(section, lines)
    lines.append("")
    lines.append(ruo_banner)
    out = "\n".join(lines)
    return unicodedata.normalize("NFC", out)


def _render_section(section: Mapping[str, Any], lines: list[str]) -> None:
    title = str(section.get("title", ""))
    rows: Sequence[Mapping[str, Any]] = section.get("rows", []) or []
    lines.append(title)
    if not rows:
        placeholder = section.get("emptyMessage") or "No findings to report."
        lines.append(f"  {placeholder}")
        return
    for row in rows:
        _render_row(row, lines)


def _render_row(row: Mapping[str, Any], lines: list[str]) -> None:
    item_id = row.get("itemId")
    value = row.get("value")
    if value is None:
        value = "Not available"
    stale = row.get("stale")
    warning = row.get("warning")
    if item_id:
        segment = row.get("segment")
        seg_str = f" (segment {segment})" if segment else ""
        lines.append(f"  - {item_id}{seg_str}: {value}")
        if stale:
            lines.append(f"    (last computed {stale.get('computedAt')})")
        if warning:
            lines.append(f"  ! {warning}")
        return
    label = row.get("label", "")
    stale_tail = f" (last computed {stale.get('computedAt')})" if stale else ""
    lines.append(f"  {label}: {value}{stale_tail}")
    if warning:
        lines.append(f"  ! {warning}")


__all__ = ["render_readout_plain_text"]

File: packages/app/src/emr/services/report/acrPlainTextRenderer.ts
// SPDX-FileCopyrightText: Copyright LiverRa
// SPDX-License-Identifier: Apache-2.0

/**
 * acrPlainTextRenderer — pure transform: ReadoutSnapshot → plain text.
 *
 * Implements contracts/plaintext-renderer.md:
 *   - RUO disclaimer first AND last line
 *   - Fixed section order
 *   - Two-space indent for fields and per-item lists
 *   - Lists prefixed `  - ` with item id + segment + summary
 *   - Warnings prefixed `  ! ` indented under the field they describe
 *   - Stale markers `(last computed <ISO time>)`
 *   - NFC unicode normalization on every output string
 *   - No markdown / HTML characters
 *   - Deterministic; no Date.now / Math.random / UUID generation
 *
 * Twin: packages/ml-inference/src/services/export/acr_plaintext_renderer.py
 * The cross-channel parity test asserts byte-equivalent output between
 * the two implementations for the shared fixture corpus.
 */

import type { ReadoutSection, ReadoutSnapshot } from './acrAnatomicalMapping';

/**
 * Convert a snapshot into clipboard-ready plain text. The string is
 * already NFC-normalized; callers MAY pass it directly to
 * `navigator.clipboard.writeText`.
 */
export function renderReadoutPlainText(snapshot: ReadoutSnapshot): string {
  const lines: string[] = [];
  const ruoBanner = `--- ${snapshot.ruoDisclaimer.replace(/^---\s*|\s*---$/g, '')} ---`;
  lines.push(ruoBanner);
  lines.push('');

  snapshot.sections.forEach((section, idx) => {
    if (idx > 0) lines.push('');
    renderSection(section, lines);
  });

  lines.push('');
  lines.push(ruoBanner);

  const out = lines.join('\n');
  return out.normalize('NFC');
}

function renderSection(section: ReadoutSection, lines: string[]): void {
  lines.push(section.title);
  if (section.rows.length === 0) {
    const placeholder = section.emptyMessage ?? 'No findings to report.';
    lines.push(`  ${placeholder}`);
    return;
  }

  // Per-item lesion rows render as `- ID (segment N): summary`. Other
  // rows render as `Label: value`.
  for (const row of section.rows) {
    if (row.itemId) {
      const segment = row.segment ? ` (segment ${row.segment})` : '';
      const value = row.value ?? 'Not available';
      lines.push(`  - ${row.itemId}${segment}: ${value}`);
      if (row.stale) {
        lines.push(`    (last computed ${row.stale.computedAt})`);
      }
      if (row.warning) {
        lines.push(`  ! ${row.warning}`);
      }
      continue;
    }
    const value = row.value ?? 'Not available';
    const staleTail = row.stale ? ` (last computed ${row.stale.computedAt})` : '';
    lines.push(`  ${row.label}: ${value}${staleTail}`);
    if (row.warning) {
      lines.push(`  ! ${row.warning}`);
    }
  }
}
